### ISSUE TITLE
[help wanted] `Invalid Macro Application` error

### ISSUE THREAD

**Issue by alex-s168** [Lambda-Mountain-Compiler-Backend/lambda-mountain#952]
I was trying to implement a (not really efficient) umbra string type like this:
 ```
type Umbra (Umbra(
  len: U32 ,
  backing: (UmbraShort(arr: Array<U8, 12>))
         | (UmbraLong(prefix: Array<U8, 4> , ptr: U8[]))
));
...
addr := λ(: s Umbra). (: (
   (match (.backing s) (
      ( (UmbraShort( arr ))
           (as (& arr) U8[]) )
      ( (UmbraLong( _ ptr )) 
           ptr )
   ))
) U8[]);
...
```

but I get this error and I don't know why:
```
Invalid Macro Application

In File PLATFORM/C/LIB/umbra.lm Line 63 Column 5
In Function Application
match-pats:
?
((uuid_000000000000584c ('UmbraShort arr)) (scope (((if (match-pats-arm (uuid_000000000000584c (as (& arr))))) 'U8[]) (scope (((if (match-pats-arm (uuid_000000000000584c ('UmbraLong (_ ptr))))) ptr) (fail ('Pattern\sMatch\sFailure_s (': ((p ':Location:) Constant + Literal + String))))))))):
?
Return:
?

```

**Comment by andrew4328** [comment]
Yeah, that error message sucks. LM doesn't support inline compound typedefs or untagged unions yet.

```
type UmbraShortLong (UmbraShort( arr:Array<U8,12> )) | (UmbraLong(prefix: Array<U8, 4> , ptr: U8[]));
type Umbra (Umbra( len:U32, backing:UmbraShortLong ));
```

or 

```
type Umbra (UmbraShort( len:U32, arr:Array<U8,12> )) | (UmbraLong( len:U32, prefix: Array<U8, 4>, ptr: U8[]));
```

**Comment by alex-s168** [comment]
I'm getting a really weird error:
`Pattern Match Failure at File: PLUGINS/BACKEND/C/compile-expr-direct.lm Line: 89 Column: 14`

[Source](http://207.180.202.42:8070/view?id=50)

**Comment by andrew4328** [comment]
Hmm, this is a nasty error. LM does not excel at optimizing custom data structures atm. Especially when interacting with the "match" destructuring primitive.

Match currently tries to dereference pointers by default to permit further destructuring or to create useful values, which is not at all what you are trying to do.

I am going to
* add an error message during typecheck for this (quick fix)
* add a ticket for making match statement lhs that permit not dereferencing (feature request)
  * I had tried this before but hit some roadblocks
* add a ticket for making codegen total (if code passes typecheck, then it should compile)
   * this can become a priority during the LSTS rewrite

**Comment by andrew4328** [comment]
field accessors directly still work `(.prefix u)` if you want to destructure manually. You can also avoid this error in match by not binding anything `(UmbraLong( _ ))`

However, I would also suggest sitting on this for a while, as what you are trying to do will be better supported in the near future. The C backend is literally about 2 weeks old.

**Comment by andrew4328** [comment]
Your code is very idiomatic. Thanks for taking the time to study so much.

**Comment by andrew4328** [comment]
a lot of good ideas coming from this
* more closer alignment with C or at least with "portable assembler" level code
* make code nearly optimal C representation 
   * like a good macro assembler should be: good like the host language but slightly more abstract or strict
 * 100% coverage of error messages 
 * full documentation of stdlib
 * onboarding of how to succeed in programming with LM (tutorials, idiomatic examples, help forums)
 * general "well-formedness"
    * internal handoffs need to be enforced
    * externally user should be guided towards correctness with correct blame and suggestions for how to correct
    * intuitive everything: nothing should be hidden, obscure, or otherwise broken beyond imagination
